giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+            memUsed + " maxMem=" + maxMem);
+        }
+        numBytesToOffload.set((long) (2 * gcReclaimFraction * maxMem) -
+          (maxMem - memUsed));
+        if (LOG.isInfoEnabled()) {
+          LOG.info("gcCompleted: tight memory usage. Starting to offload " +
+            "until " + numBytesToOffload.get() + " bytes are offloaded");
+        }
+        state = State.OFFLOADING;
+        updateRates(1, 1);
+      }
+    }
+  }
+
+  /**
+   * Given an estimate for the current memory usage and the maximum available
+   * memory, it updates the active threads and flow control credit in the
+   * OOC engine.
+   *
+   * @param usageEstimateMem Estimate of memory usage.
+   * @param maxMemory Maximum memory.
+   */
+  private void updateRates(long usageEstimateMem, long maxMemory) {
+    double usageEstimate = (double) usageEstimateMem / maxMemory;
+    if (usageEstimate > 0) {
+      if (usageEstimate >= amHighThreshold) {
+        oocEngine.updateActiveThreadsFraction(0);
+      } else if (usageEstimate < amLowThreshold) {
+        oocEngine.updateActiveThreadsFraction(1);
+      } else {
+        oocEngine.updateActiveThreadsFraction(1 -
+          (usageEstimate - amLowThreshold) /
+            (amHighThreshold - amLowThreshold));
+      }
+
+      if (usageEstimate >= creditHighThreshold) {
+        oocEngine.updateRequestsCreditFraction(0);
+      } else if (usageEstimate < creditLowThreshold) {
+        oocEngine.updateRequestsCreditFraction(1);
+      } else {
+        oocEngine.updateRequestsCreditFraction(1 -
+          (usageEstimate - creditLowThreshold) /
+            (creditHighThreshold - creditLowThreshold));
+      }
+    } else {
+      oocEngine.updateActiveThreadsFraction(1);
+      oocEngine.updateRequestsCreditFraction(1);
+    }
+  }
+
+  /**
+   * Returns statistics about the old gen pool.
+   * @return {@link MemoryUsage}.
+   */
+  private MemoryUsage getOldGenUsed() {
+    List<MemoryPoolMXBean> memoryPoolList =
+      ManagementFactory.getMemoryPoolMXBeans();
+    for (MemoryPoolMXBean pool : memoryPoolList) {
+      String normalName = pool.getName().toLowerCase();
+      if (normalName.contains("old") || normalName.contains("tenured")) {
+        return pool.getUsage();
+      }
+    }
+    throw new IllegalStateException("Bad Memory Pool");
+  }
+
+  /**
+   * Maintains statistics about the current state and progress of the
+   * computation and produces estimates of memory usage using a technique
+   * based on linear regression.
+   *
+   * Upon a GC events, it gets updated with the most recent statistics through
+   * the {@link #addRecord} method.
+   */
+  private static class MemoryEstimator {
+    /** Stores the (x1,x2,...,x5) arrays of data samples, one for each sample */
+    private Vector<double[]> dataSamples = new Vector<>();
+    /** Stores the y memory usage dataSamples, one for each sample */
+    private Vector<Double> memorySamples = new Vector<>();
```

Comment:
`DoubleArrayList` from fastutil is useful when we have a lot of entry in the list. The number of entries in our case is very small (the number of full GCs in one superstep, which is in the range of 1-100, getting closer to 1000 in the very extreme cases). So, the choice of either `Vector` or `ArrayList` is more natural. I agree that we are not using the synchronized guarantees of `Vector`, hence using `ArrayList` is an even more natural choice. I'll change these to `ArrayList`. Conversion from `ArrayList` (or even `Vector`) to primitive array can be done through `toArray` method in either of the collections. I'll fix that too.